### PR TITLE
One click release of quarantined mail directly from quarantine report

### PR DESCRIPTION
**Pull request by asuweb**
This pull request provides simple one-click release of quarantined messages directly from the daily quarantine report.

This is an often sought after feature that will greatly improve an admins life, and provide a much more streamlined end user experience.

Minimal changes have been made to the base code.

### CODE REVIEW

**Comment by shawniverson** [comment]
There needs to be an expiry period for entries in mod_release so that mod_release does not continuously grow when users do not release messages.

**Comment by asuweb** [comment]
@shawniverson - agreed.  I intend to tie this into the existing quarantine expiry, so that when the built in mailwatch cleanup runs, it'll also clean the mod_release table.

**Comment by asuweb** [comment]
Expiry on mod_release now working.  There was a typo in the quarantine_maint.php file where id was used instead of msg_id for the mod_release table.  

Records in mod_release will now be deleted when removed from the quarantine.

File: mailscanner/auto-release.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * Requires: Mailwatch 1.2.0
+ *
+ * Provides the mechanism for one click release of quarantined emails as reported by the quarantine_report.php cron
+ *
+ * SETUP:
+ *
+ * Create the following table in the mailscanner database:
+ * CREATE TABLE `mod_release` (
+ *  `id` int(11) NOT NULL AUTO_INCREMENT,
```
Comment (endelwar) [inline mailscanner/auto-release.php]: Use bigint ad id type to help installation on high volume systems

File: mailscanner/auto-release.php [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+ * Enter database credentials below.
+ */
+if (isset($_GET['mid']) && isset($_GET['r'])) {
+    // Change the following to reflect the location of functions.php
+    require_once('/var/www/html/mailscanner/functions.php');
+    //Database Credentials
+    $host = 'localhost';//change if using a remote db
+    $user = '';
+    $pass = '';
+    $database = 'mailscanner'; //change is your database is called something else
```
Comment (endelwar) [inline mailscanner/auto-release.php]: Host, user, password and dbname should be read from conf.php

File: mailscanner/auto-release.php [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    $db = mysqli_connect($host, $user, $pass, $database) or die("CONNECT ERROR" . mysqli_connect_error());
+    $mid = mysqli_real_escape_string($db, $_GET['mid']);
+    $token = mysqli_real_escape_string($db, $_GET['r']);
+    $query = "SELECT * FROM mod_release WHERE msg_id = '$mid'";
+    $result = mysqli_query($db, $query);
```
Comment (endelwar) [inline mailscanner/auto-release.php]: I know that it sounds anacronistic in 2016, bit alla mailwatch 1.2 code uses MySQL estention...

File: mailscanner/auto-release.php [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    if (!$result) {
+        die("Error fetching from database");
+    }
+    if (mysqli_num_rows($result) == 0) {
+        echo "<p>Message not found.  You may have already released this message.</p>
+<p>Please contact your email administrator and provide them with this message ID: ".$mid." if you need this message released</p> ";
```
Comment (endelwar) [inline mailscanner/auto-release.php]: Please sue the translation function for string output

File: tools/Cron_jobs/quarantine_maint.php [hunk 5]
```diff
@@ -84,6 +84,11 @@ function quarantine_clean()
                         // Update the quarantine flag
                         $sql = "UPDATE maillog SET timestamp=timestamp, quarantined = NULL WHERE id='$id'";
                         dbquery($sql);
+                        //ASU MODIFIED CODE BEGIN
+                        //remove any corresponding record in mod_release
+                        $sql = "DELETE FROM mod_release WHERE msg_id = '$id'";
```
Comment (endelwar) [inline tools/Cron_jobs/quarantine_maint.php]: Add a flag in conf.PHP and  an if nere to bypass this query if autorelease is not enabled

File: tools/Cron_jobs/quarantine_report.php [hunk 6]
```diff
@@ -315,6 +315,27 @@ function return_quarantine_list_array($to_address, $to_domain)
     return $array;
 }
 
+//ASU MODIFIED CODE BEGIN
+function get_random_string()
+{
+    $bytes = openssl_random_pseudo_bytes(10);
```
Comment (endelwar) [inline tools/Cron_jobs/quarantine_report.php]: Set a $count variabile as a parameter to function

File: tools/Cron_jobs/quarantine_report.php [hunk 7]
```diff
@@ -315,6 +315,27 @@ function return_quarantine_list_array($to_address, $to_domain)
...
+function store_auto_release($qitem)
+{
+    $id = $qitem['id'];
+    $rand = $qitem['rand'];
+    $result = dbquery("INSERT INTO mod_release (msg_id,uid) VALUES ('$id','$rand')");
+    if (!$result) {
+        dbg(" ==== Error generating auto_release....skipping...");
+        return false;
+    } else {
+        return true;
```
Comment (endelwar) [inline tools/Cron_jobs/quarantine_report.php]: This should return $rand, that has to be used in release link

File: tools/Cron_jobs/quarantine_report.php [hunk 8]
```diff
@@ -323,6 +344,15 @@ function send_quarantine_email($email, $filter, $quarantined)
     $t1 = "";
     // Build the quarantine list for this recipient
     foreach ($quarantined as $qitem) {
+        //ASU MODIFIED CODE BEGIN
+        $qitem['rand'] = get_random_string();
+        $auto_release = store_auto_release($qitem);
+        if ($auto_release) {
+            $links = '<a href="' . QUARANTINE_REPORT_HOSTURL . '/viewmail.php?id=' . $qitem['id'] . '">View</a>  <a href="' . QUARANTINE_REPORT_HOSTURL . '/auto-release.php?mid=' . $qitem['id'] . '&r=' . $qitem['rand'] . '">Release</a>';
+        } else {
+            $links = '<a href="' . QUARANTINE_REPORT_HOSTURL . '/viewmail.php?id=' . $qitem['id'] . '">View</a>';
```
Comment (endelwar) [inline tools/Cron_jobs/quarantine_report.php]: See note on line 334

File: tools/Cron_jobs/quarantine_report.php (hunk 7 above)
Comment (asuweb) [inline tools/Cron_jobs/quarantine_report.php]: The generated random value is stored in $qitem['rand'], for use in the email link

**Comment by asuweb** [comment]
Closed to allow branch change - ref #260

**Pull request closed without merging**